File: examples/go/internal/rs/rs.go
// Package rs is the minimal Resource Server used by the demo.
//
// In gateway mode (what this demo uses), Keycloak validates the agent+jwt,
// runs constraint checks, and then proxies the request to this server. The
// server simply executes the business logic and returns a JSON body
// wrapped in a `data` field (per the spec's §5.11 sync response shape).
//
// A production resource server running in **direct mode** would also
// validate the agent JWT itself — either by calling Keycloak's
// /agent/introspect or by checking signature + aud locally against the
// agent's registered public key. For this mini demo we rely on gateway
// mode so the resource server has nothing to auth.
package rs

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
)

// CapabilityHandler turns parsed arguments into a JSON-serializable data
// object. The server wraps the result in {"data": <handler output>}.
type CapabilityHandler func(args map[string]any) map[string]any

// ResourceServer is a running HTTP server keyed by path → handler.
type ResourceServer struct {
	Server   *http.Server
	Listener net.Listener
	Port     int
}

// StartOn binds 0.0.0.0:port and begins serving; returns once the listener
// is accepting. Port 0 picks a random free port — handy for tests.
func StartOn(port int, handlers map[string]CapabilityHandler) (*ResourceServer, error) {
	mux := http.NewServeMux()
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		handle(w, r, handlers)
	})

	ln, err := net.Listen("tcp", fmt.Sprintf("0.0.0.0:%d", port))
	if err != nil {
		return nil, err
	}
	boundPort := ln.Addr().(*net.TCPAddr).Port

	srv := &http.Server{Handler: mux}
	log.Printf("[resource-server] listening on :%d", boundPort)
	go func() {
		if err := srv.Serve(ln); err != nil && err != http.ErrServerClosed {
			log.Printf("[resource-server] serve error: %v", err)
		}
	}()
	return &ResourceServer{Server: srv, Listener: ln, Port: boundPort}, nil
}

// Close stops the server.
func (s *ResourceServer) Close() error {
	return s.Server.Close()
}

// URLFor builds the capability URL Keycloak (and the rest of the compose
// network) uses to reach this server.
func (s *ResourceServer) URLFor(path, serviceHost string) string {
	if serviceHost == "" {
		serviceHost = "resource-server"
	}
	return fmt.Sprintf("http://%s:%d%s", serviceHost, s.Port, path)
}

func handle(w http.ResponseWriter, r *http.Request, handlers map[string]CapabilityHandler) {
	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	handler, ok := handlers[r.URL.Path]
	if !ok {
		w.WriteHeader(http.StatusNotFound)
		_ = json.NewEncoder(w).Encode(map[string]string{
			"error": "not_found",
			"path":  r.URL.Path,
		})
		return
	}
	bodyBytes, err := io.ReadAll(r.Body)
	if err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		_ = json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
		return
	}
	var parsed map[string]any
	if len(bodyBytes) > 0 {
		if err := json.Unmarshal(bodyBytes, &parsed); err != nil {
			w.WriteHeader(http.StatusBadRequest)
			_ = json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
			return
		}
	}
	args, _ := parsed["arguments"].(map[string]any)
	if args == nil {
		args = map[string]any{}
	}
	argsJSON, _ := json.Marshal(args)
	log.Printf("[resource-server] %s  args=%s", r.URL.Path, string(argsJSON))

	data := handler(args)
	w.Header().Set("content-type", "application/json")
	w.WriteHeader(http.StatusOK)
	_ = json.NewEncoder(w).Encode(map[string]any{"data": data})
}

File: examples/rust/src/rs.rs
//! The **Resource Server** core — where capability business logic runs.
//!
//! In gateway mode (what this demo uses), Keycloak validates the agent+jwt,
//! runs constraint checks, then proxies to this server. We just execute
//! and return `{"data": {...}}` per AAP §5.11 sync response shape.

use std::collections::HashMap;
use std::convert::Infallible;
use std::net::SocketAddr;
use std::sync::Arc;

use anyhow::Result;
use bytes::Bytes;
use http_body_util::{BodyExt, Full};
use hyper::body::Incoming;
use hyper::server::conn::http1;
use hyper::service::service_fn;
use hyper::{Method, Request, Response, StatusCode};
use hyper_util::rt::TokioIo;
use serde_json::{json, Value};
use tokio::net::TcpListener;

pub type Handler = Box<dyn Fn(&Value) -> Value + Send + Sync>;

pub struct ResourceServer {
    handlers: Arc<HashMap<String, Handler>>,
}

impl ResourceServer {
    pub fn new(handlers: HashMap<String, Handler>) -> Self {
        Self {
            handlers: Arc::new(handlers),
        }
    }

    /// Bind on `0.0.0.0:port` and serve forever.
    pub async fn serve(self, port: u16) -> Result<()> {
        let addr: SocketAddr = ([0, 0, 0, 0], port).into();
        let listener = TcpListener::bind(addr).await?;
        println!("[resource-server] listening on :{port}");

        loop {
            let (stream, _) = listener.accept().await?;
            let handlers = self.handlers.clone();
            tokio::spawn(async move {
                let svc = service_fn(move |req| handle(req, handlers.clone()));
                if let Err(e) = http1::Builder::new()
                    .serve_connection(TokioIo::new(stream), svc)
                    .await
                {
                    eprintln!("[resource-server] connection error: {e}");
                }
            });
        }
    }
}

async fn handle(
    req: Request<Incoming>,
    handlers: Arc<HashMap<String, Handler>>,
) -> Result<Response<Full<Bytes>>, Infallible> {
    if req.method() != Method::POST {
        return Ok(empty(StatusCode::METHOD_NOT_ALLOWED));
    }
    let path = req.uri().path().to_string();
    let handler = match handlers.get(&path) {
        Some(h) => h,
        None => {
            return Ok(json_response(
                StatusCode::NOT_FOUND,
                &json!({"error": "not_found", "path": path}),
            ));
        }
    };

    let body_bytes = match req.into_body().collect().await {
        Ok(c) => c.to_bytes(),
        Err(_) => return Ok(empty(StatusCode::BAD_REQUEST)),
    };
    let parsed: Value = if body_bytes.is_empty() {
        Value::Null
    } else {
        match serde_json::from_slice(&body_bytes) {
            Ok(v) => v,
            Err(_) => return Ok(empty(StatusCode::BAD_REQUEST)),
        }
    };
    let args = parsed
        .get("arguments")
        .cloned()
        .unwrap_or(Value::Object(Default::default()));
    println!(
        "[resource-server] {path}  args={}",
        serde_json::to_string(&args).unwrap_or_default()
    );
    let data = handler(&args);
    Ok(json_response(StatusCode::OK, &json!({"data": data})))
}

fn empty(status: StatusCode) -> Response<Full<Bytes>> {
    Response::builder()
        .status(status)
        .body(Full::new(Bytes::new()))
        .unwrap()
}

fn json_response(status: StatusCode, body: &Value) -> Response<Full<Bytes>> {
    let bytes = serde_json::to_vec(body).unwrap_or_default();
    Response::builder()
        .status(status)
        .header("content-type", "application/json")
        .body(Full::new(Bytes::from(bytes)))
        .unwrap()
}
